File: c++ version/1-100/92-reverse-linked-list-II.cpp
ListNode* reverseBetween(ListNode* head, int left, int right) {
    ListNode* dummy = new ListNode(-1);
    dummy->next = head;
    ListNode* pre = dummy;
    for (int i = 0; i < left-1; ++i) {
        pre = pre->next;
    }
    ListNode* start = pre->next;
    for (int j = 0; j < right-left; ++j) {
        ListNode* tmp = pre->next;
        pre->next = start->next;
        start->next = start->next->next;
        pre->next->next = tmp;
    }

    ListNode* newHead = dummy->next;
    delete dummy;
    return newHead;
}

File: python version/92-reverse-linked-list-II.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseLink(self, head):
        prev = None
        curr = head
        while curr:
            tmp = curr.next
            curr.next = prev
            prev = curr
            curr = tmp   
        return prev

    def reverseBetween(self, head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
        if left == 1: 
            prev = None
            reverse_start = head
        else:
            prev = head
            for i in range(left-2):
                prev = prev.next
            reverse_start = prev.next
        curr = reverse_start
        for i in range(right-left):
            curr = curr.next
        reverse_end = curr
        nxt = reverse_end.next
        reverse_end.next = None
        if prev:
            prev.next = self.reverseLink(reverse_start)
        else:
            head = self.reverseLink(reverse_start)
        
        curr = head
        while curr.next:
            curr = curr.next
        curr.next = nxt

        return head
        
            